mmaction/utils/gradcam_utils.py

```diff
@@ -0,0 +1,188 @@
+        else:
+            # [batch_size, num_crops*num_clips, 3, clip_len, H, W]
+            b1, b2, _, T, H, W = inputs['imgs'].size()
+            B = b1 * b2
+
+        gradients = self.target_gradients
+        activations = self.target_activations
+        if self.is_recognizer2d:
+            # [B*Tg, C', H', W']
+            BB, C, _, _ = gradients.size()
+            Tg = BB // B
+        else:
+            # source shape: [B, Tg, C', H', W']
+            _, C, Tg, _, _ = gradients.size()
+            # target shape: [B, C', Tg, H', W']
+            gradients = gradients.permute(0, 2, 1, 3, 4)
+            activations = activations.permute(0, 2, 1, 3, 4)
+
+        # calculate & resize to [B, 1, T, H, W]
+        weights = torch.mean(gradients.view(B, Tg, C, -1), dim=3)
+        weights = weights.view(B, Tg, C, 1, 1)
+        activations = activations.view([B, Tg, C] +
+                                       list(activations.size()[-2:]))
+        localization_map = torch.sum(
+            weights * activations, dim=2, keepdim=True)
+        localization_map = F.relu(localization_map)
+        localization_map = localization_map.permute(0, 2, 1, 3, 4)
+        localization_map = F.interpolate(
+            localization_map,
+            size=(T, H, W),
+            mode='trilinear',
+            align_corners=False,
+        )
+
+        # Normalize the localization map.
+        localization_map_min, localization_map_max = (
+            torch.min(localization_map.view(B, -1), dim=-1, keepdim=True)[0],
+            torch.max(localization_map.view(B, -1), dim=-1, keepdim=True)[0],
+        )
+        localization_map_min = torch.reshape(
+            localization_map_min, shape=(B, 1, 1, 1, 1))
+        localization_map_max = torch.reshape(
+            localization_map_max, shape=(B, 1, 1, 1, 1))
+        localization_map = (localization_map - localization_map_min) / (
+            localization_map_max - localization_map_min + 1e-12)
+        localization_map = localization_map.data
+
+        return localization_map.squeeze(dim=1), preds
+
+    def __call__(self, inputs, use_labels=False, alpha=0.5):
+        """Visualize the localization maps on their corresponding inputs as
+        heatmap, using Grad-CAM.
+
+        Args:
+            inputs (dict): model inputs, generated by test pipeline.
+            use_labels (bool): Whether to use given labels to generate
+                localization map. Labels are in `inputs['label']`.
+            alpha (float): transparency level of the heatmap,
+                in the range [0, 1].
+        Returns:
+        """
+        imgs = inputs['imgs']
+
+        # localization_map shape [B, T, H, W]
+        # 2D: B = batch_size, T = num_segments
+        # 3D: B = batch_size * num_crops * num_clips, T = clip_len
+        localization_map, preds = self._calculate_localization_map(
+            inputs, use_labels=use_labels)
+        if localization_map.device != torch.device('cpu'):
```

Comment:
Seems that localization_map should eventually be on cpu, maybe just avoid this if-branch for better unittest coverage.